src/Data/Compactable.purs

```diff
@@ -0,0 +1,156 @@
+module Data.Compactable
+  ( class Compactable
+  , compact
+  , separate
+  , compactDefault
+  , separateDefault
+  ) where
+
+import Control.Applicative (class Applicative, class Apply, apply, pure)
+import Control.Bind (class Bind, bind, join)
+import Data.Array as Array
+import Data.Either (Either(Right, Left), hush, note)
+import Data.Foldable (class Foldable, foldl)
+import Data.Function (($))
+import Data.Functor (class Functor, map, (<$>))
+import Data.List as List
+import Data.Map as Map
+import Data.Maybe (Maybe(..))
+import Data.Monoid (class Monoid, mempty, (<>))
+import Data.Traversable (class Traversable, sequence, traverse)
+import Data.Tuple (Tuple(..))
+import Prelude (class Ord, unit, (<<<))
+
+class Compactable f where
+  compact :: forall a. f (Maybe a) -> f a
+  separate :: forall l r. f (Either l r) -> { left :: f l, right :: f r }
+
+compactDefault :: forall f a. Functor f => Compactable f => f (Maybe a) -> f a
+compactDefault = _.right <<< separate <<< map (note unit)
+
+separateDefault
+  :: forall f l r
+   . Functor f
+  => Compactable f
+  => f (Either l r) -> { left :: f l, right :: f r}
+separateDefault xs = { left: compact $ (hush <<< swapEither) <$> xs
+                     , right: compact $ hush <$> xs
+                     }
+  where
+    swapEither e = case e of
+      Left x  -> Right x
+      Right y -> Left y
+
+instance compactableMaybe :: Compactable Maybe where
+  compact = join
+
+  separate Nothing = { left: Nothing, right: Nothing }
+  separate (Just e) = case e of
+    Left l  -> { left: Just l, right: Nothing }
+    Right r -> { left: Nothing, right: Just r }
+
+instance compactableEither :: Monoid m => Compactable (Either m) where
+  compact (Left m) = Left m
+  compact (Right m) = case m of
+    Just v  -> Right v
+    Nothing -> Left mempty
+
+  separate (Left x) = { left: Left x, right: Left x }
+  separate (Right e) = case e of
+    Left l  -> { left: Right l, right: Left mempty }
+    Right r -> { left: Left mempty, right: Right r }
+
+instance compactableArray :: Compactable Array where
+  compact = Array.catMaybes
+  separate xs = separateSequence xs
+
+instance compactableList :: Compactable List.List where
+  compact = List.catMaybes
+  separate xs = separateSequence xs
+
+separateSequence
+  :: forall f l r
+   . Monoid (f l)
+  => Monoid (f r)
+  => Foldable f
+  => Applicative f
+  => Compactable f
+  => f (Either l r) -> { left :: f l, right :: f r }
+separateSequence = foldl go { left: mempty, right: mempty } where
+    go acc e = case e of
+      Left l  -> acc { left = acc.left <> pure l }
+      Right r -> acc { right = acc.right <> pure r }
```

Comment:
probably want `Alternative f` here, which will give you `pure`, `empty`, and `<|>` (which are guaranteed to work for all `l` and `r`), instead of the monoid and applicative instances.

could maybe give your accumulator a newtype and monoid instance given `Alternative f`, use `foldMap`, but that’s not necessary.

could also use this for `Compatable Maybe` ;)